engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala

```diff
@@ -0,0 +1,60 @@
+package pl.touk.nussknacker.engine.flink.util.transformer.aggregate.median
+
+import pl.touk.nussknacker.engine.api.typed.supertype.NumberTypesPromotionStrategy.{
+  ForLargeFloatingNumbersOperation,
+  ForMathOperation
+}
+import pl.touk.nussknacker.engine.util.MathUtils
+
+import javax.annotation.Nullable
+import scala.annotation.tailrec
+import scala.util.Random
+
+object MedianHelper {
+  private val rand = new Random(42)
+
+  @Nullable
+  def calculateMedian(numbers: List[Number]): Number = {
+    if (numbers.isEmpty) {
+      null
+    } else if (numbers.size % 2 == 1) {
+      MathUtils.convertToPromotedType(quickSelect(numbers, (numbers.size - 1) / 2))(ForLargeFloatingNumbersOperation)
+    } else {
+      // it is possible to fetch both numbers with single recursion, but it would complicate code
+      val firstNumber  = quickSelect(numbers, numbers.size / 2 - 1)
+      val secondNumber = quickSelect(numbers, numbers.size / 2)
+
+      val sum = MathUtils.largeFloatingSum(firstNumber, secondNumber)
+      MathUtils.divideWithDefaultBigDecimalScale(sum, 2)
+    }
+  }
+
+  @tailrec
+  private def quickSelect(numbers: List[Number], indexToTake: Int): Number = {
+    require(numbers.nonEmpty)
+    require(indexToTake >= 0)
+    require(indexToTake < numbers.size)
+
+    val randomElement = numbers(rand.nextInt(numbers.size))
+    val groupedBy = numbers.groupBy(e => {
+      val cmp = MathUtils.compare(e, randomElement)
+      if (cmp < 0) {
```

Comment:
in the comparable interface compareTo, (I assume that this method has similar contract), it is documented that it returns negative if first is lower. So it can return value other than those three values. Should I move this if to the MathUtils.compare implementation?

Reply:
You're right. There is Math.signum but it is defined only for float and double.